src/replication.c

```diff
@@ -3223,15 +3250,53 @@ void setupMainConnForPsync(connection *conn) {
                                      "accept connections in read-write mode.\n");
         }
         dualChannelSyncHandlePsync();
-        return;
+        return C_OK;
     }
+    *err = getTryPsyncString(psync_result);
+    return C_ERR;
+}
 
-error:
+/* Replication: Replica side.
+ * This connection handler fires after rdb-connection was initialized. We use it
+ * to adjust the replica main for loading incremental changes into the local buffer. */
+void dualChannelSetupMainConnForPsync(connection *conn) {
+    char *err = NULL;
+    int ret;
+
+    switch (server.repl_state) {
+    case REPL_STATE_SEND_HANDSHAKE:
+        ret = dualChannelReplMainConnSendHandshake(conn, &err);
+        server.repl_state = REPL_STATE_RECEIVE_CAPA_REPLY;
+        break;
+    case REPL_STATE_RECEIVE_CAPA_REPLY:
+        ret = dualChannelReplMainConnRecvCapaReply(conn, &err);
+        if (ret == C_ERR) {
+            break;
+        }
```

Comment:
Right, `cancelReplicationHandshake` will not clean correctly when `dualChannelReplMainConnRecvPsyncReply` fails. We should not change state upon C_ERR.

This case is unique because we are actually trying to perform two replication activities within the same callback call (recive capability reply and send a psync request).